sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala

```diff
@@ -95,6 +96,29 @@ abstract class LogicalPlan extends QueryPlan[LogicalPlan] with Logging {
   }
 
   /**
+   * Returns the default statistics or statistics estimated by cbo based on configuration.
+   */
+  final def planStats(conf: CatalystConf): Statistics = {
+    if (conf.cboEnabled) {
+      if (estimatedStats.isEmpty) {
+        estimatedStats = Some(cboStatistics(conf))
+      }
+      estimatedStats.get
+    } else {
+      statistics
+    }
+  }
+
+  /**
+   * Returns statistics estimated by cbo. If the plan doesn't override this, it returns the
+   * default statistics.
+   */
+  protected def cboStatistics(conf: CatalystConf): Statistics = statistics
+
+  /** A cache for the estimated statistics, such that it will only be computed once. */
+  private var estimatedStats: Option[Statistics] = None
```

Comment:
The conf won't change in a query, but if we cache plans across queries, then we can't cache the  calculated stats.  And now I do find such plans e.g. `cachedDataSourceTables`.
If we don't cache the stats, estimation can be a performance hit.
Then we need to use lazy val and don't depend on the conf. In the current stage it's ok. I just thought in the future we may need some parameters in more complicated estimation. e.g. we may need a threshold to determine whether a column is a primary key.